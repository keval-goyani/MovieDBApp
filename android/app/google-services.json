{
  "project_info": {
    "project_number": "89996564819",
    "project_id": "moviedb-2908d",
    "storage_bucket": "moviedb-2908d.appspot.com"
  },
  "client": [
    {
      "client_info": {
        "mobilesdk_app_id": "1:89996564819:android:e942630a87714b1d3ea9a6",
        "android_client_info": {
          "package_name": "com.moviedbapp"
        }
      },
      "oauth_client": [
        {
          "client_id": "89996564819-hph4f8e5b765gev43v87vj04t9tun4il.apps.googleusercontent.com",
          "client_type": 3
        }
      ],
      "api_key": [
        {
          "current_key": "AIzaSyBYrM1MC9Go35l8ra748zqBtCRDkrB_Nn4"
        }
      ],
      "services": {
        "appinvite_service": {
          "other_platform_oauth_client": [
            {
              "client_id": "89996564819-hph4f8e5b765gev43v87vj04t9tun4il.apps.googleusercontent.com",
              "client_type": 3
            }
          ]
        }
      }
    }
  ],
  "configuration_version": "1"
}