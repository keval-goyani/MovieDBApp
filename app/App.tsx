import React, {useEffect} from 'react';
import {SafeAreaView, Text} from 'react-native';
import SplashScreen from 'react-native-splash-screen';
import {styles} from './theme';

const App = () => {
  useEffect(() => {
    SplashScreen.hide();
  }, []);

  return (
    <SafeAreaView style={styles.container}>
      <Text>Hello World</Text>
    </SafeAreaView>
  );
};

export default App;
