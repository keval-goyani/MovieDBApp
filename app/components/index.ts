import ChatHeader from './ChatHeader';
import ChatInput from './ChatInput';
import CustomSkeleton from './CustomSkeleton';
import DropDownMenu from './DropDownMenu';
import Header from './Header';
import List from './List';
import ListContainer, { movieListItem } from './ListContainer';
import Loader from './Loader';
import LoadingState from './LoadingState';
import Message from './Message';
import MessageList from './MessageList';
import Movies from './Movies';
import MovieTrailer, { trailerListItem } from './MovieTrailers';
import SearchModal from './SearchModal';
import SkeletonCard from './SkeletonCard';
import TabBarIcon from './TabBarIcon';
import Trailers from './Trailers';
import UserListEmpty from './UserListEmpty';
import UsersList from './UsersList';

export {
  ChatHeader,
  ChatInput,
  DropDownMenu,
  Header,
  List,
  ListContainer,
  MovieTrailer,
  Message,
  MessageList,
  TabBarIcon,
  SearchModal,
  Loader,
  LoadingState,
  Movies,
  movieListItem,
  Trailers,
  trailerListItem,
  UsersList,
  SkeletonCard,
  CustomSkeleton,
  UserListEmpty,
};
