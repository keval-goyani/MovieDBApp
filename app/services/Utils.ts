import firestore from '@react-native-firebase/firestore';
import storage from '@react-native-firebase/storage';
import apisauce from 'apisauce';
import React, { Dispatch } from 'react';
import { Alert } from 'react-native';
import documentPicker from 'react-native-document-picker';
import FileViewer from 'react-native-file-viewer';
import {
  Asset,
  launchCamera,
  launchImageLibrary,
} from 'react-native-image-picker';
import {
  check,
  openSettings,
  PERMISSIONS,
  request,
  RESULTS,
} from 'react-native-permissions';
import RNFetchBlob from 'rn-fetch-blob';
import { ImmutableObject } from 'seamless-immutable';
import {
  appConstants,
  ChatDataType,
  ClearChatDataType,
  DetailResponseGenerator,
  DocumentStateDataType,
  genres,
  ListItemDataType,
  MovieDetailsDataType,
  MovieResponseGenerator,
  pickerOptions,
  strings,
} from '../constants';
import { Color, Metrics } from '../theme';

export const apiConfig = apisauce.create({
  baseURL: appConstants.baseUrl,
});

export async function getError(
  response: MovieResponseGenerator | DetailResponseGenerator,
) {
  if (response.problem === strings.clientError) {
    return strings.pageErrorMessage;
  }
  if (response?.problem === strings.networkError) {
    return strings.networkErrorMessage;
  }
  if (
    [strings.connectionError, strings.serverError].includes(response?.problem)
  ) {
    return strings.serverErrorMessage;
  }
  return strings.errorMessage;
}

export const alertMessage = (error: string) => {
  Alert.alert(error);
};

export const getDetails = (
  movieDetails: ImmutableObject<MovieDetailsDataType> | null,
) => {
  const time = movieDetails?.runtime ?? movieDetails?.episode_run_time[0] ?? 0;
  const votePercentage = (movieDetails?.vote_average ?? 0) * 10;
  const activeStrokeColor =
    votePercentage > 69
      ? Color.PercentageDarkGreen
      : Color.percentageDarkYellow;
  const inActiveStrokeColor =
    votePercentage > 69
      ? Color.PercentageLightGreen
      : Color.percentageLightYellow;
  const year = new Date(
    movieDetails?.release_date ?? movieDetails?.first_air_date ?? '',
  )
    .toString()
    .slice(11, 15);
  const country =
    movieDetails?.production_countries.length === 0
      ? ''
      : `(${movieDetails?.production_countries?.[0]?.iso_3166_1})`;
  const movieTitle = movieDetails?.title ?? movieDetails?.name;
  const hour = Math.floor(time / 60);
  const minutes = time % 60;
  const runTime =
    minutes === 0
      ? `${hour}h`
      : hour === 0
      ? `${minutes}m`
      : `${hour}h ${minutes}m`;
  const movieType = movieDetails?.genres?.map(item => item.name).join(', ');
  const directorName = movieDetails?.credits?.crew
    .filter(item => item.job === strings.director)
    .map(item => item.name)
    .join(', ');

  return {
    votePercentage,
    activeStrokeColor,
    inActiveStrokeColor,
    year,
    country,
    movieTitle,
    runTime,
    movieType,
    directorName,
  };
};

const getGenreID = (query: string) => {
  return genres.filter(
    genre => genre?.name.toLowerCase() === query.toLowerCase(),
  )[0]?.id;
};

const searchConditionCheck = (item: string, query: string) => {
  return item?.toLowerCase().includes(query.toLowerCase());
};

export const searchMovie = (movieList: ListItemDataType[], query: string) => {
  const genreId = getGenreID(query);

  return movieList.filter(
    movie =>
      searchDatacheck(movie, query) || movie.genre_ids?.includes(genreId),
  );
};

const searchDatacheck = (
  {
    title,
    original_title,
    original_name,
    name = '',
    overview,
  }: ListItemDataType,
  query: string,
) => {
  return (
    searchConditionCheck(title, query) ||
    searchConditionCheck(original_title, query) ||
    searchConditionCheck(original_name, query) ||
    searchConditionCheck(name, query) ||
    searchConditionCheck(overview, query)
  );
};

export const getUniqueMovies = (movies: ListItemDataType[]) => {
  const uniqueMovies = [
    ...new Map(movies.map(item => [item.id, item])).values(),
  ];

  return uniqueMovies;
};

const padTo2Digits = (input: number) => {
  return input.toString().padStart(2, '0');
};

export const timestampToTime = (timestamp: number) => {
  const hours = new Date(timestamp).getHours();
  const minutes = new Date(timestamp).getMinutes();

  return hours >= 12
    ? `${padTo2Digits(hours - 12)}:${padTo2Digits(minutes)} pm`
    : `${padTo2Digits(hours)}:${padTo2Digits(minutes)} am`;
};

export const getChatTime = (time: number) => {
  const latestChatTime = new Date(time).getDate();
  const currentTime = new Date().getDate();
  const timeDifference = currentTime - latestChatTime;

  if (time !== 0) {
    switch (timeDifference) {
      case 0:
        return timestampToTime(time);
      case 1:
        return strings.yesterday;
      default:
        return new Date(time).toLocaleDateString('en-GB').toString();
    }
  }
  return strings.emptyString;
};

export const sortString = (input: string) => {
  return input.split('').sort().join('');
};

export const chatIdCreation = (userId: string, id: string) => {
  return userId.localeCompare(id) > 0 ? userId + id : id + userId;
};

export const handleCameraPermission = (
  setImagePath: Dispatch<React.SetStateAction<string>>,
) => {
  check(appConstants.cameraPermission)
    .then(result => {
      switch (result) {
        case RESULTS.BLOCKED:
          permissionAlert(strings.cameraPermission);
          break;
        case RESULTS.DENIED:
          requestCameraPermission();
          break;
        case RESULTS.UNAVAILABLE || RESULTS.LIMITED:
          alertMessage(strings.permissionUnavailable);
          break;
        default:
          openCamera(setImagePath);
      }
    })
    .catch(error => {
      alertMessage(error.message);
    });
};

export const handleGalleryPermission = (
  setImagePath: Dispatch<React.SetStateAction<string>>,
) => {
  check(appConstants.galleryPermission)
    .then(result => {
      switch (result) {
        case RESULTS.BLOCKED:
          permissionAlert(strings.galleryPermission);
          break;
        case RESULTS.DENIED:
          requestGalleryPermission();
          break;
        case RESULTS.UNAVAILABLE || RESULTS.LIMITED:
          alertMessage(strings.permissionUnavailable);
          break;
        default:
          selectImage(setImagePath);
      }
    })
    .catch(error => {
      alertMessage(error.message);
    });
};

export const handleDocumentPermission = (
  setDocumentData: Dispatch<React.SetStateAction<DocumentStateDataType>>,
) => {
  if (Metrics.isAndroid) {
    check(appConstants.documentWritePermission).then(result => {
      switch (result) {
        case RESULTS.BLOCKED:
          permissionAlert(strings.document);
          break;
        case RESULTS.DENIED:
          request(PERMISSIONS.ANDROID.WRITE_EXTERNAL_STORAGE);
          break;
        case RESULTS.UNAVAILABLE || RESULTS.LIMITED:
          alertMessage(strings.permissionUnavailable);
          break;
        default:
          selectDocument(setDocumentData);
      }
    });
  } else {
    selectDocument(setDocumentData);
  }
};

const selectDocument = async (
  setDocumentData: Dispatch<React.SetStateAction<DocumentStateDataType>>,
) => {
  await documentPicker
    .pickSingle({
      type: [
        documentPicker.types.pdf,
        documentPicker.types.images,
        documentPicker.types.pptx,
        documentPicker.types.ppt,
        documentPicker.types.doc,
        documentPicker.types.docx,
        documentPicker.types.csv,
        documentPicker.types.zip,
      ],
      presentationStyle: 'fullScreen',
      copyTo: 'documentDirectory',
    })
    .then(response => {
      addDocumentToStorage(
        response?.fileCopyUri ?? '',
        response.name,
        setDocumentData,
      );
    })
    .catch(error => error);
};

const addDocumentToStorage = async (
  url: string,
  documentName: string,
  setDocumentData: Dispatch<React.SetStateAction<DocumentStateDataType>>,
) => {
  const documentTitle = `${appConstants.timestamp}_${documentName}`;
  const documentPath = `${appConstants.storageDocumentPath}${documentTitle}`;

  await storage()
    .ref(documentPath)
    .putFile(url)
    .then(response => {
      const stroredDocumentPath = Metrics.isAndroid
        ? `${appConstants.storageDocumentPath}${response.metadata.name}`
        : `${response.metadata.name}`;

      storage()
        .ref(stroredDocumentPath)
        .getDownloadURL()
        .then(remoteDocumentUrl => {
          setDocumentData({
            documentUrl: remoteDocumentUrl,
            documentName: documentTitle,
          });
        });
    })
    .catch(error => alertMessage(error.message));
};

const requestCameraPermission = () => {
  return Metrics.isAndroid
    ? request(PERMISSIONS.ANDROID.CAMERA)
    : request(PERMISSIONS.IOS.CAMERA);
};

const requestGalleryPermission = () => {
  return Metrics.isAndroid
    ? request(PERMISSIONS.ANDROID.WRITE_EXTERNAL_STORAGE)
    : request(PERMISSIONS.IOS.PHOTO_LIBRARY_ADD_ONLY);
};

const permissionAlert = (type: string) => {
  const permissionBlockedAlert =
    type === strings.cameraPermission
      ? strings.cameraPermissionBlocked
      : strings.galleryPermissionBlocked;

  Alert.alert(permissionBlockedAlert, strings.grantPermission, [
    { onPress: () => openSettings(), text: strings.goToSettings },
    { text: strings.cancel },
  ]);
};

const pickerCallback = (
  assets: Asset[] = [],
  setImagePath: Dispatch<React.SetStateAction<string>>,
) => {
  const { androidFolder, iosFolder } = appConstants;
  const filePath = generateFilePath(assets[0]?.fileName);
  const data = assets[0]?.base64 ?? '';
  const directory = Metrics.isAndroid ? androidFolder : iosFolder;

  RNFetchBlob.fs.isDir(directory).then(isDir => {
    if (isDir) {
      filePath && saveImage(filePath, data, setImagePath);
    } else {
      RNFetchBlob.fs
        .mkdir(directory)
        .then(() => filePath && saveImage(filePath, data, setImagePath))
        .catch(error => alertMessage(error.message));
    }
  });
};

const generateFilePath = (fileName: string = '') => {
  const { androidFolder, iosFolder, timestamp } = appConstants;

  if (fileName) {
    return Metrics.isAndroid
      ? `${androidFolder}/${timestamp}_${fileName}`
      : `${iosFolder}/${timestamp}_${fileName}`;
  }
};

const saveImage = (
  path: string,
  data: string,
  setImagePath: Dispatch<React.SetStateAction<string>>,
) => {
  RNFetchBlob.fs
    .createFile(path, data, 'base64')
    .then(() => getImage(path, setImagePath))
    .catch(error => alertMessage(error.message));
};

const getImage = (
  path: string,
  setImagePath: Dispatch<React.SetStateAction<string>>,
) => {
  RNFetchBlob.fs
    .readStream(path, 'base64')
    .then(({ path: imagePath }) => {
      setImagePath(imagePath);
    })
    .catch(error => alertMessage(error.message));
};

const openCamera = (setImagePath: Dispatch<React.SetStateAction<string>>) => {
  launchCamera({ mediaType: 'photo', ...pickerOptions })
    .then(
      ({ assets, didCancel }) =>
        !didCancel && pickerCallback(assets, setImagePath),
    )
    .catch(error => alertMessage(error.message));
};

const selectImage = (setImagePath: Dispatch<React.SetStateAction<string>>) => {
  launchImageLibrary({ mediaType: 'photo', ...pickerOptions })
    .then(
      ({ assets, didCancel }) =>
        !didCancel && pickerCallback(assets, setImagePath),
    )
    .catch(error => alertMessage(error.message));
};

export const clearChat = ({
  navigation,
  chatId,
  setShowMenu,
}: ClearChatDataType) => {
  Alert.alert(strings.areYouSure, strings.clearChatAlert, [
    { text: strings.cancel, onPress: () => setShowMenu(false) },
    {
      text: strings.ok,
      onPress: () => {
        firestore().collection(strings.chatCollection).doc(chatId).delete();
        setShowMenu(false);
        navigation.goBack();
      },
    },
  ]);
};

export const chatCreation = async (
  chatId: string,
  uid: string,
  content: string,
  type: string,
  documentData: DocumentStateDataType,
  setMessageList: Dispatch<React.SetStateAction<ChatDataType[]>>,
) => {
  const timeStamp = Date.now();
  let data = {};
  if (type === strings.document) {
    const { documentUrl, documentName } = documentData;
    data = {
      content: documentUrl,
      type,
      user: uid,
      time: timeStamp,
      documentName,
    };
  } else {
    data = {
      content,
      type,
      user: uid,
      time: timeStamp,
    };
  }

  const previousMessage = await firestore()
    .collection(strings.chatCollection)
    .doc(chatId)
    .get()
    .then(documentSnapshot => documentSnapshot.data());

  setMessageList([...(previousMessage?.messageList ?? ''), data]);
};

export const openDocument = (url: string, documentName: string) => {
  const filePath = Metrics.isAndroid
    ? `${appConstants.androidDocumentFolder}/${documentName}`
    : `${appConstants.iosDocumentFolder}/${documentName}`;
  const folderPath = Metrics.isAndroid
    ? appConstants.androidDocumentFolder
    : appConstants.iosDocumentFolder;

  RNFetchBlob.fs.isDir(folderPath).then(isDir => {
    if (!isDir) {
      RNFetchBlob.fs.mkdir(folderPath);
    }
  });

  RNFetchBlob.fs.exists(filePath).then(response => {
    response
      ? FileViewer.open(filePath)
          .then(openFileResponse => openFileResponse)
          .catch(() => alertMessage(strings.appCheck))
      : RNFetchBlob.config({
          addAndroidDownloads: {
            useDownloadManager: true,
            notification: false,
            path: filePath,
          },
          path: filePath,
        })
          .fetch('GET', url)
          .then(documentPathResponse => {
            FileViewer.open(documentPathResponse?.data)
              .then(openFileResponse => openFileResponse)
              .catch(() => alertMessage(strings.appCheck));
          })
          .catch(error => {
            alertMessage(error);
          });
  });
};

export const addChatToFirestore = async (
  chatId: string,
  messageList: ChatDataType[],
) => {
  messageList.length !== 0 &&
    (await firestore()
      .collection(strings.chatCollection)
      .doc(chatId)
      .set({ messageList })
      .catch(error => error));
};
