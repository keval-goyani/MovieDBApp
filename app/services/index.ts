import { getPopularMovieData } from './Api';
import immutablePersistenceTransform from './immutablePersistenceTransform';
import {
  alertMessage,
  apiConfig,
  loginError,
  chatCreation,
  clearChat,
  conversationIdCreation,
  convertToTimestamp,
  decryptData,
  encryptData,
  getChatTime,
  getDetails,
  getError,
  getUniqueMovies,
  handleCameraPermission,
  handleDocumentPermission,
  handleGalleryPermission,
  openDocument,
  searchMovie,
  signUpError,
  sortString,
  timestampToTime,
  alertBox,
  storeImageToStorage,
  groupCreation,
  dateConvertor,
  clearConversation,
  sub,
  handleLogOut,
  chatImageHandler,
} from './Utils';

export {
  immutablePersistenceTransform,
  alertMessage,
  apiConfig,
  getDetails,
  getError,
  getPopularMovieData,
  searchMovie,
  getUniqueMovies,
  getChatTime,
  timestampToTime,
  sortString,
  handleGalleryPermission,
  handleCameraPermission,
  clearChat,
  chatCreation,
  handleDocumentPermission,
  conversationIdCreation,
  openDocument,
  encryptData,
  decryptData,
  convertToTimestamp,
  loginError,
  signUpError,
  alertBox,
  storeImageToStorage,
  groupCreation,
  dateConvertor,
  clearConversation,
  sub,
  handleLogOut,
  chatImageHandler,
};
