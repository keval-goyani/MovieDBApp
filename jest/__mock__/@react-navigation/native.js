export const navigation = {
  navigate: jest.fn(),
  goBack: jest.fn(),
  replace: jest.fn(),
  popToTop: jest.fn(),
  addListener: jest.fn().mockImplementation((str, cb) => cb()),
  pop: jest.fn(),
  getParam: jest.fn(),
  push: jest.fn(),
};

const useRoute = jest.fn(val => val);
// const useFocusEffect = jest.fn();
const useIsFocused = jest.fn(val => val);
const useNavigation = () => navigation;
const setFocus = jest.fn();
const createNavigationContainerRef = jest.fn();
const DefaultTheme = {};

const CommonActions = { reset: jest.fn() };
const StackActions = {
  pop: jest.fn(),
};

export {
  useRoute,
  useNavigation,
  // useFocusEffect,
  useIsFocused,
  setFocus,
  createNavigationContainerRef,
  DefaultTheme,
  CommonActions,
  StackActions,
};
