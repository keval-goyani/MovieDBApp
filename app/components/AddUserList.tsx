import { useFocusEffect, useNavigation } from '@react-navigation/native';
import React, { useEffect, useState } from 'react';
import { FlatList, Text, TouchableOpacity, View } from 'react-native';
import Check from 'react-native-vector-icons/AntDesign';
import { useDispatch, useSelector } from 'react-redux';
import { asMutable } from 'seamless-immutable';
import { ProfileImage, UserListEmpty } from '../components';
import {
  AddUserListProps,
  NavigationDataType,
  navigationStrings,
  RenderItemTypes,
  SelectedUsersProps,
  strings,
  UserDataType,
} from '../constants';
import { authDataSelectors } from '../redux/AuthRedux';
import { userListSelector } from '../redux/UserListRedux';
import { conversationIdCreation, sub } from '../services';
import { Color, moderateScale } from '../theme';
import { styles } from './styles/AddUserListStyles';

const AddUsersList = ({ userListData, setSelectedUsers }: AddUserListProps) => {
  const navigation: NavigationDataType = useNavigation();
  const dispatch = useDispatch();
  const { user } = useSelector(authDataSelectors.getData);
  const { fetchingUserList } = useSelector(userListSelector.getData);
  const [selectedItems, setSelectedItems] = useState<UserDataType[]>([]);

  useFocusEffect(() => {
    sub({ user, dispatch });
    return () => sub;
  });

  useEffect(() => {
    setSelectedUsers(selectedItems);
  }, [selectedItems, setSelectedUsers]);

  const renderUserList = ({ item, index }: RenderItemTypes) => {
    const usersEmail = [user?.email ?? '', item?.email];
    const conversationId: string = conversationIdCreation(usersEmail);

    const handleOnPress = (selectedUser: UserDataType) => {
      console.log(selectedItems, '<==selectedItems.length');
      console.log(selectedUser, '<==selectedUser');

      if (selectedItems.length) {
        return selectUsers(selectedUser);
      }
      navigation.goBack();
      navigation.navigate(navigationStrings.Chat, {
        conversationId,
        username: item?.username,
        receiverId: item?.uid,
        userStatus: item?.status,
        profileImage: item?.profileImage,
        userEmail: item?.email,
      });
    };

    const getSelected = (usersData: SelectedUsersProps) => {
      return JSON.stringify(selectedItems).includes(JSON.stringify(usersData));
    };

    const selected = getSelected(item);

    const selectUsers = (selectedUser: UserDataType) => {
      if (
        JSON.stringify(selectedItems).includes(JSON.stringify(selectedUser))
      ) {
        const filteredList = selectedItems?.filter(
          (items: UserDataType) => items !== selectedUser,
        );
        setSelectedItems(filteredList);
      } else {
        setSelectedItems([...selectedItems, selectedUser]);
      }
    };

    return (
      <TouchableOpacity
        style={styles.listItem}
        onPress={() => handleOnPress(item)}x
        activeOpacity={0.8}
        onLongPress={() => selectUsers(item)}
        testID={`button-for-test${index}`}>
        <View style={styles.itemContainer}>
          <ProfileImage
            profileImage={item?.profileImage}
            userStatus={item?.status}
          />
          <View style={styles.nameView}>
            <Text style={styles.text}>{item?.username}</Text>
            <Text
              style={styles.lastChatText}
              ellipsizeMode="tail"
              numberOfLines={1}>
              {item?.email}
            </Text>
          </View>
        </View>
        {selected && (
          <>
            <View style={styles.overlay} />
            <Check
              testID="check"
              name={strings.checkIcon}
              size={moderateScale(23)}
              color={Color.lightBlue}
              style={styles.iconStyle}
            />
          </>
        )}
      </TouchableOpacity>
    );
  };

  return (
    <FlatList
      testID={'flatList'}
      data={asMutable(userListData)}
      renderItem={renderUserList}
      ListEmptyComponent={
        <UserListEmpty
          fetching={fetchingUserList}
          userListLength={userListData.length}
        />
      }
      showsVerticalScrollIndicator={false}
      bounces={false}
    />
  );
};

export default AddUsersList;
