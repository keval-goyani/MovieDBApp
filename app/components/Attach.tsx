import { useNavigation } from '@react-navigation/native';
import React, { FC } from 'react';
import { Alert, View } from 'react-native';
import SystemSetting from 'react-native-system-setting';
import { CustomIconRounder } from '../components';
import {
  AttachDataType,
  NavigationDataType,
  navigationStrings,
  strings,
} from '../constants';
import {
  handleCameraPermission,
  handleDocumentPermission,
  handleGalleryPermission,
} from '../services';
import { Icons } from '../theme';
import styles from './styles/AttachStyles';

const Attach: FC<AttachDataType> = ({
  setIsAttach,
  setImagePath,
  setDocumentData,
  conversationId,
  username,
  receiverId,
}) => {
  const navigation: NavigationDataType = useNavigation();

  const turnOnLocation = (enable: boolean) => {
    if (enable) {
      navigation.navigate(navigationStrings.Location, {
        isFromChat: false,
        ...{ conversationId, username, receiverId },
      });
      setIsAttach(false);
    } else {
      Alert.alert(strings.turnOn, strings.goToSetting, [
        {
          onPress: () => SystemSetting.switchLocation(),
          text: strings.setting,
        },
        { text: strings.cancel },
      ]);
    }
  };

  const locationHandler = () => {
    SystemSetting.isLocationEnabled()
      .then(enable => {
        turnOnLocation(enable);
      })
      .catch(error => Alert.alert(strings.warning, error));
  };

  return (
    <View style={styles.container}>
      <View style={styles.iconContainer}>
        <CustomIconRounder
          backgroundColor={styles.documentBackgroundColor}
          path={Icons.fileIcon}
          iconName={strings.document}
          tintColor={styles.documentTintColor}
          onPress={() => {
            setIsAttach(false);
            handleDocumentPermission(setDocumentData);
          }}
        />
        <CustomIconRounder
          backgroundColor={styles.cameraBackgroundColor}
          path={Icons.cameraIcon}
          iconName={strings.camera}
          tintColor={styles.cameraTintColor}
          onPress={() => {
            handleCameraPermission(setImagePath);
            setIsAttach(false);
          }}
        />
        <CustomIconRounder
          backgroundColor={styles.galleryBackgroundColor}
          path={Icons.galleryIcon}
          iconName={strings.gallery}
          tintColor={styles.galleryTintColor}
          onPress={() => {
            handleGalleryPermission(setImagePath);
            setIsAttach(false);
          }}
        />
        <CustomIconRounder
          backgroundColor={styles.locationBackgroundColor}
          path={Icons.locationIcon}
          iconName={strings.map}
          tintColor={styles.locationTintColor}
          onPress={locationHandler}
        />
      </View>
    </View>
  );
};

export default Attach;
