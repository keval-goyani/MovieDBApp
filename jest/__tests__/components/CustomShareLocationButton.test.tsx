import { navigation } from '@react-navigation/native';
import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import { useSelector } from 'react-redux';
import { CustomShareLocationButton } from '../../../app/components';
import * as Services from '../../../app/services/Utils';

describe('Custom Share Location Button', () => {
  const props = {
    isFromChat: false,
    conversationId: 'C12345678',
    latitude: 32.345678,
    longitude: 34.65748,
    receiverId: 'R123444',
  };

  const userData = {
    uid: 'U1234556',
  };

  it('should take snapshot of customShareLocstionButton', () => {
    const { toJSON } = render(<CustomShareLocationButton {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('shpuld render the share location button when isFromChat is false', () => {
    const { getByTestId } = render(<CustomShareLocationButton {...props} />);
    const shareLocationButton = getByTestId('share-location-button');
    expect(shareLocationButton).toBeDefined();
  });

  it('should call shareLocationHandler without senderId and receiverId', () => {
    const spy = jest.spyOn(Services, 'chatCreation');
    const customProps = {
      isFromChat: false,
      conversationId: 'C12345678',
      latitude: 32.345678,
      longitude: 34.65748,
    };

    const { getByTestId } = render(
      <CustomShareLocationButton {...customProps} />,
    );
    const shareLocationButton = getByTestId('share-location-button');
    fireEvent.press(shareLocationButton);
    expect(spy.mock.calls[0][0].senderId).toEqual('');
    expect(spy.mock.calls[0][0].receiverId).toEqual('');
  });

  it('should call shareLocationHandler and  navigate.goBack and call the chatCreation function when shareLocationButton is pressed', async () => {
    const spy = jest.spyOn(Services, 'chatCreation').mockImplementation();

    useSelector.mockReturnValue({
      user: userData,
    });

    const locationMessage = {
      conversationId: props.conversationId,
      senderId: userData.uid,
      receiverId: props.receiverId,
      content: `${props.latitude},${props.longitude}`,
      type: 'Location',
    };

    const { getByTestId } = render(<CustomShareLocationButton {...props} />);
    const shareLocationButton = getByTestId('share-location-button');
    fireEvent.press(shareLocationButton);

    await expect(spy).toHaveBeenCalledWith(locationMessage);
    expect(navigation.goBack).toHaveBeenCalled();
  });
});
