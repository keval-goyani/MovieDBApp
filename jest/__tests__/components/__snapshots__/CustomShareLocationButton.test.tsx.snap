// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Share Location Button should take snapshot of customShareLocstionButton 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#0d253f",
      "padding": undefined,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "opacity": 1,
      }
    }
    testID="share-location-button"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#0edbb0",
          "borderRadius": 30,
          "height": undefined,
          "justifyContent": "center",
          "width": undefined,
        }
      }
    >
      <Image
        source={
          Object {
            "default": Object {
              "attach": Object {
                "testUri": "../../../app/assets/icons/attach.png",
              },
              "avatar": Object {
                "testUri": "../../../app/assets/icons/user.png",
              },
              "backIcon": Object {
                "testUri": "../../../app/assets/icons/backIcon.png",
              },
              "camera": Object {
                "testUri": "../../../app/assets/icons/camera.png",
              },
              "cameraIcon": Object {
                "testUri": "../../../app/assets/icons/cameraIcon.png",
              },
              "chatBackground": Object {
                "testUri": "../../../app/assets/icons/chatBackground.png",
              },
              "close": Object {
                "testUri": "../../../app/assets/icons/close.png",
              },
              "comingSoon": Object {
                "testUri": "../../../app/assets/icons/comingSoon.png",
              },
              "communityIcon": Object {
                "testUri": "../../../app/assets/icons/communityIcon.png",
              },
              "document": Object {
                "testUri": "../../../app/assets/icons/documentIcon.png",
              },
              "documentIcon": Object {
                "testUri": "../../../app/assets/icons/documentIcon.png",
              },
              "dotMenu": Object {
                "testUri": "../../../app/assets/icons/dots.png",
              },
              "expandIcon": Object {
                "testUri": "../../../app/assets/icons/expandIcon.png",
              },
              "fileIcon": Object {
                "testUri": "../../../app/assets/icons/fileIcon.png",
              },
              "fullStop": Object {
                "testUri": "../../../app/assets/icons/fullStop.png",
              },
              "galleryIcon": Object {
                "testUri": "../../../app/assets/icons/galleryIcon.png",
              },
              "homeIcon": Object {
                "testUri": "../../../app/assets/icons/homeIcon.png",
              },
              "imageDocument": Object {
                "testUri": "../../../app/assets/icons/imageIcon.png",
              },
              "location": Object {
                "testUri": "../../../app/assets/icons/share.png",
              },
              "locationIcon": Object {
                "testUri": "../../../app/assets/icons/locationIcon.png",
              },
              "logout": Object {
                "testUri": "../../../app/assets/icons/logout.png",
              },
              "menuIcon": Object {
                "testUri": "../../../app/assets/icons/menuIcon.png",
              },
              "movieDbIcon": Object {
                "testUri": "../../../app/assets/icons/movieDBIcon.png",
              },
              "movieDbIcon2x": Object {
                "testUri": "../../../app/assets/icons/movieDB2x.png",
              },
              "movieDbIcon3x": Object {
                "testUri": "../../../app/assets/icons/movieDB3x.png",
              },
              "moviesIcon": Object {
                "testUri": "../../../app/assets/icons/moviesIcon.png",
              },
              "noItemFound": Object {
                "testUri": "../../../app/assets/icons/noItemFound.png",
              },
              "notFound": Object {
                "testUri": "../../../app/assets/icons/notFoundIcon.png",
              },
              "pdfDocument": Object {
                "testUri": "../../../app/assets/icons/pdfDocument 2.png",
              },
              "playIcon": Object {
                "testUri": "../../../app/assets/icons/playIcon.png",
              },
              "plus": Object {
                "testUri": "../../../app/assets/icons/plusSign.png",
              },
              "pptDocument": Object {
                "testUri": "../../../app/assets/icons/pptDocument.png",
              },
              "searchIcon": Object {
                "testUri": "../../../app/assets/icons/searchIcon.png",
              },
              "send": Object {
                "testUri": "../../../app/assets/icons/send.png",
              },
              "threeDotIcon": Object {
                "testUri": "../../../app/assets/icons/threeDotIcon.png",
              },
              "trailerIcon": Object {
                "testUri": "../../../app/assets/icons/trailerIcon.png",
              },
              "trending": Object {
                "testUri": "../../../app/assets/icons/trending.png",
              },
              "tv": Object {
                "testUri": "../../../app/assets/icons/tv.png",
              },
              "videoDocument": Object {
                "testUri": "../../../app/assets/icons/mp4Document.png",
              },
            },
          }
        }
        style={
          Object {
            "height": undefined,
            "tintColor": "#0d253f",
            "width": undefined,
          }
        }
      />
    </View>
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontSize": undefined,
          "fontWeight": "500",
          "paddingLeft": undefined,
        }
      }
    >
      Share live location
    </Text>
  </View>
</View>
`;
